modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/io/exporter/DDMFormInstanceRecordWriterTrackerImpl.java

```diff
@@ -51,50 +52,82 @@ public Map<String, String> getDDMFormInstanceRecordWriterExtensions() {
 			_ddmFormInstanceRecordWriterExtensions);
 	}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY
-	)
-	protected void addDDMFormInstanceRecordWriter(
-		DDMFormInstanceRecordWriter ddmFormInstanceRecordWriter,
-		Map<String, Object> properties) {
-
-		String type = MapUtil.getString(
-			properties, "ddm.form.instance.record.writer.type");
-
-		String extension = MapUtil.getString(
-			properties, "ddm.form.instance.record.writer.extension");
-
-		if (Validator.isNull(extension)) {
-			extension = StringUtil.toUpperCase(type);
-		}
-
-		_ddmFormInstanceRecordWriterExtensions.put(type, extension);
-
-		_ddmFormInstanceRecordWriters.put(type, ddmFormInstanceRecordWriter);
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTracker = ServiceTrackerFactory.open(
+			bundleContext, DDMFormInstanceRecordWriter.class,
+			new DDMFormInstanceRecordWriterServiceTrackerCustomizer(
+				bundleContext));
 	}
 
 	@Deactivate
 	protected void deactivate() {
-		_ddmFormInstanceRecordWriterExtensions.clear();
+		_serviceTracker.close();
 		_ddmFormInstanceRecordWriters.clear();
-	}
-
-	protected void removeDDMFormInstanceRecordWriter(
-		DDMFormInstanceRecordWriter ddmFormInstanceRecordWriter,
-		Map<String, Object> properties) {
-
-		String type = MapUtil.getString(
-			properties, "ddm.form.instance.record.writer.type");
-
-		_ddmFormInstanceRecordWriterExtensions.remove(type);
-		_ddmFormInstanceRecordWriters.remove(type);
+		_ddmFormInstanceRecordWriterExtensions.clear();
 	}
 
 	private final Map<String, String> _ddmFormInstanceRecordWriterExtensions =
 		new TreeMap<>();
 	private final Map<String, DDMFormInstanceRecordWriter>
 		_ddmFormInstanceRecordWriters = new TreeMap<>();
+	private volatile ServiceTracker
+		<DDMFormInstanceRecordWriter, DDMFormInstanceRecordWriter>
+			_serviceTracker;
+
+	private class DDMFormInstanceRecordWriterServiceTrackerCustomizer
+		implements ServiceTrackerCustomizer
+			<DDMFormInstanceRecordWriter, DDMFormInstanceRecordWriter> {
+
+		public DDMFormInstanceRecordWriterServiceTrackerCustomizer(
+			BundleContext bundleContext) {
+
+			_bundleContext = bundleContext;
+		}
+
+		@Override
+		public DDMFormInstanceRecordWriter addingService(
+			ServiceReference<DDMFormInstanceRecordWriter> serviceReference) {
+
+			String type = (String)serviceReference.getProperty(
+				"ddm.form.instance.record.writer.type");
+
+			String extension = (String)serviceReference.getProperty(
+				"ddm.form.instance.record.writer.extension");
+
+			if (Validator.isNull(extension)) {
+				extension = StringUtil.toUpperCase(type);
+			}
+
+			_ddmFormInstanceRecordWriterExtensions.put(type, extension);
+			_ddmFormInstanceRecordWriters.put(
+				type, _bundleContext.getService(serviceReference));
```

Comment:
@JanisZhang Please avoid `_bundleContext.getService` twice. Use a variable to hold the service, put it into the map, and return it.

Reply:
Got it, thanks Dante!